gateway/shard/manager.go

```diff
@@ -0,0 +1,314 @@
+package shard
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/diamondburned/arikawa/v3/api"
+	"github.com/diamondburned/arikawa/v3/discord"
+	"github.com/diamondburned/arikawa/v3/gateway"
+	"github.com/diamondburned/arikawa/v3/internal/backoff"
+	"github.com/pkg/errors"
+)
+
+func updateIdentifier(ctx context.Context, id *gateway.Identifier) (url string, err error) {
+	botData, err := api.NewClient(id.Token).WithContext(ctx).BotURL()
+	if err != nil {
+		return "", err
+	}
+
+	if botData.Shards < 1 {
+		botData.Shards = 1
+	}
+
+	id.Shard = &gateway.Shard{0, botData.Shards}
+
+	// Update the burst to be the current given time and reset it back to
+	// the default when the given time is reached.
+	id.IdentifyGlobalLimit.SetBurst(botData.StartLimit.Remaining)
+	resetAt := time.Now().Add(botData.StartLimit.ResetAfter.Duration())
+	id.IdentifyGlobalLimit.SetBurstAt(resetAt, botData.StartLimit.Total)
+
+	// Update the maximum number of identify requests allowed per 5s.
+	id.IdentifyShortLimit.SetBurst(botData.StartLimit.MaxConcurrency)
+
+	return botData.URL, nil
+}
+
+// Manager is the manager responsible for handling all sharding on this
+// instance. An instance of Manager must never be copied.
+type Manager struct {
+	// shards are the *shards.shards managed by this Manager. They are
+	// sorted in ascending order by their shard id.
+	shards     []shard
+	gatewayURL string
+
+	mutex sync.RWMutex
+
+	rescaling *rescalingState // nil unless rescaling
+
+	new NewShardFunc
+}
+
+type rescalingState struct {
+	haltRescale context.CancelFunc
+	rescaleDone sync.WaitGroup
+}
+
+// NewManager creates a Manager using as many gateways as recommended by
+// Discord.
+func NewManager(token string, new NewShardFunc) (*Manager, error) {
+	id := gateway.DefaultIdentifier(token)
+
+	url, err := updateIdentifier(context.Background(), id)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get gateway info")
+	}
+
+	return NewIdentifiedManagerWithURL(url, id, new)
```

Comment:
`new` is a built-in function so, even though it's not used in the function, we should opt for another name.
```suggestion
func NewManager(token string, fn NewShardFunc) (*Manager, error) {
	id := gateway.DefaultIdentifier(token)

	url, err := updateIdentifier(context.Background(), id)
	if err != nil {
		return nil, errors.Wrap(err, "failed to get gateway info")
	}

	return NewIdentifiedManagerWithURL(url, id, fn)
```